ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/DLModel.java

```diff
@@ -0,0 +1,307 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.opensearch.ml.engine.ModelHelper.ONNX_ENGINE;
+import static org.opensearch.ml.engine.ModelHelper.ONNX_FILE_EXTENSION;
+import static org.opensearch.ml.engine.ModelHelper.PYTORCH_ENGINE;
+import static org.opensearch.ml.engine.ModelHelper.PYTORCH_FILE_EXTENSION;
+import static org.opensearch.ml.engine.utils.FileUtils.deleteFileQuietly;
+
+@Log4j2
+public abstract class DLModel implements Predictable {
+    public static final String MODEL_ZIP_FILE = "model_zip_file";
+    public static final String MODEL_HELPER = "model_helper";
+    public static final String ML_ENGINE = "ml_engine";
+    protected ModelHelper modelHelper;
+    protected MLEngine mlEngine;
+    protected String modelId;
+
+    protected Predictor<Input, Output>[] predictors;
+    protected ZooModel[] models;
+    protected Device[] devices;
+    protected AtomicInteger nextDevice = new AtomicInteger(0);
+
+    @Override
+    public MLOutput predict(MLInput mlInput, MLModel model) {
+        throw new MLException("model not loaded");
+    }
+
+    @Override
+    public MLOutput predict(MLInput mlInput) {
+        if (modelHelper == null || modelId == null) {
+            throw new MLException("model not loaded");
+        }
+        try {
+            return AccessController.doPrivileged((PrivilegedExceptionAction<ModelTensorOutput>) () -> {
+                Thread.currentThread().setContextClassLoader(getClass().getClassLoader());
+                if (predictors == null) {
+                    throw new MLException("model not loaded.");
+                }
+                return predict(modelId, mlInput.getInputDataset());
+            });
+        } catch (Throwable e) {
+            String errorMsg = "Failed to inference " + mlInput.getAlgorithm() + " model: " + modelId;
+            log.error(errorMsg, e);
+            throw new MLException(errorMsg, e);
+        }
+    }
+
+    protected Predictor<Input, Output> getPredictor() {
+        int currentDevice = nextDevice.getAndIncrement();
+        if (currentDevice > devices.length - 1) {
+            currentDevice = currentDevice % devices.length;
+            nextDevice.set(currentDevice + 1);
+        }
+        return predictors[currentDevice];
+    }
+
+    public abstract ModelTensorOutput predict(String modelId, MLInputDataset inputDataSet) throws TranslateException;
+
+    @Override
+    public void initModel(MLModel model, Map<String, Object> params) {
+        String engine;
+        switch (model.getModelFormat()) {
+            case TORCH_SCRIPT:
+                engine = PYTORCH_ENGINE;
+                break;
+            case ONNX:
+                engine = ONNX_ENGINE;
+                break;
+            default:
+                throw new IllegalArgumentException("unsupported engine");
+        }
+
+        File modelZipFile = (File)params.get(MODEL_ZIP_FILE);
```

Comment:
This model only support initialization from a local zip file in the ml-common host? I think it should also download file from url and load?

Reply:
Download from URL is a step before this part. That means model already downloaded to local before this line.